internal/runners/shim/shim.go

```diff
@@ -39,35 +45,50 @@ func New(prime primeable) *Shim {
 	}
 }
 
-func (s *Shim) Run(args ...string) error {
-	project, fail := project.GetSafe()
-	if fail != nil {
-		// Do not fail if we can't find the projectfile
-		logging.Debug("Project not found, error: %v", fail)
-	}
-
-	if project != nil && !subshell.IsActivated() {
-		runtime, err := runtime.NewRuntime(s.proj.Source().Path(), s.proj.CommitUUID(), s.proj.Owner(), s.proj.Name(), runbits.NewRuntimeMessageHandler(s.out))
-		if err != nil {
-			return locale.WrapError(err, "err_shim_runtime_init", "Could not initialize runtime for shim command.")
-		}
-		venv := virtualenvironment.New(runtime)
-		if fail := venv.Activate(); fail != nil {
-			logging.Errorf("Unable to activate state: %s", fail.Error())
-			return locale.WrapError(fail.ToError(), "err_shim_activate", "Could not activate environment for shim command")
-		}
+func NewParams() *Params {
+	return &Params{}
+}
 
-		env, err := venv.GetEnv(true, filepath.Dir(projectfile.Get().Path()))
-		if err != nil {
-			return err
+func (s *Shim) Run(params *Params, args ...string) error {
+	if params.Path != "" {
+		var fail *failures.Failure
+		s.proj, fail = project.FromPath(params.Path)
+		if fail != nil {
+			return locale.WrapInputError(fail.ToError(), "shim_no_project_at_path", "Could not find project file at {{.V0}}", params.Path)
 		}
-		s.subshell.SetEnv(env)
+	}
+	if s.proj == nil {
+		return locale.NewError("shim_no_project_found", "Could not find a project.  You need to be in a project directory or specify a global default project via `state activate --default`")
 	}
 
 	if len(args) == 0 {
 		return nil
 	}
 
+	runtime, err := runtime.NewRuntime(s.proj.Source().Path(), s.proj.CommitUUID(), s.proj.Owner(), s.proj.Name(), runbits.NewRuntimeMessageHandler(s.out))
+	if err != nil {
+		return locale.WrapError(err, "err_shim_runtime_init", "Could not initialize runtime for shim command.")
+	}
+	venv := virtualenvironment.New(runtime)
+	if fail := venv.Activate(); fail != nil {
+		logging.Errorf("Unable to activate state: %s", fail.Error())
+		return locale.WrapError(fail.ToError(), "err_shim_activate", "Could not activate environment for shim command")
+	}
+
+	env, err := venv.GetEnv(true, filepath.Dir(projectfile.Get().Path()))
+	if err != nil {
+		return err
+	}
+	logging.Debug("Trying to shim %s on PATH=%s", args[0], env["PATH"])
+	// Ensure that we are not calling the shim recursively
+	oldval, ok := env[constants.ShimEnvVarName]
+	if ok && oldval == args[0] {
+		return locale.NewError("err_shim_recursive_loop", "Could not resolve shimmed executable {{.V0}}", args[0])
+	}
+	env[constants.ShimEnvVarName] = args[0]
```

Comment:
Seems we can drop this

Reply:
I would advocate for leaving it in.  I can still think of cases, where we could run into this issue eg.,

1. Project at path `p` provides executable `e`
1. User runs `state activate --default` => `e` is shimmed at `<bindir>/e`
1. User updates project by removing the artifact that provides `e`
1. User calls `e` resolving to `<bindir>/e` => using project path `p`, but the actual executable `e` is removed => we try `<bindir>/e`

That is only one case that I could think of right now.   There may be others, and this check is fairly simple and effective.